### PR TITLE
KAPA library prep

### PR DESCRIPTION
**Pull request by EGCGDanForth**
New scripts to support the KAPA Library Prep workflow.
-calculating UDF values
-generating input files for Hamilton
-autoplacement of samples

### CODE REVIEW

File: EPPs/common.py [hunk 1]
```diff
@@ -173,6 +183,105 @@ def patch_entry(cls, *args, **kwargs):
...
+    def write_csv(filename, csv_array):
+        """Write the list of list to the file provided as a csv file"""
+        with open(filename, 'w') as csvFile:
+            writer = csv.writer(csvFile)
+            writer.writerows(csv_array)
+
+    @cached_property
+    def input_container_names(self):
+        """The name of containers from input artifacts"""
+        containers = []
```
Comment (mwhamgenomics) [inline EPPs/common.py]: Seems a bit roundabout to build a list, then cast it to a set, then back to a list - would be simpler to just build a set and cast it at the end. I also don't see the need for `frozenset` here

File: tests/test_autoplacement_24_in_96.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        # per input
+        with self.patched_process1, self.patched_lims:
+            self.epp._run()
+
+            expected_output_placement = [
+                (fake_outputs_per_input[0], (fake_selected_containers[0], 'A:1')),
+                (fake_outputs_per_input[0], (fake_selected_containers[0], 'B:1'))
+            ]
+            self.epp.process.step.set_placements.assert_called_with(fake_selected_containers, expected_output_placement)
+            # assert self.epp.process.step.get_placement_list == ['']
```
Comment (mwhamgenomics) [inline tests/test_autoplacement_24_in_96.py]: What do we want to do with this commented assert?

File: tests/test_autoplacement_qpcr_384.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                (fake_outputs_per_input[2], (fake_selected_containers[0], 'K:5')),
+                (fake_outputs_per_input[0], (fake_selected_containers[0], 'M:1')),
+                (fake_outputs_per_input[1], (fake_selected_containers[0], 'M:3')),
+                (fake_outputs_per_input[2], (fake_selected_containers[0], 'M:5')),
+                (fake_outputs_per_input[0], (fake_selected_containers[0], 'B:1')),
+                (fake_outputs_per_input[1], (fake_selected_containers[0], 'A:2')),
+                (fake_outputs_per_input[2], (fake_selected_containers[0], 'B:2'))
+            ]
+            self.epp.process.step.set_placements.assert_called_with(fake_selected_containers, expected_output_placement)
+            # assert self.epp.process.step.get_placement_list == ['']
```
Comment (mwhamgenomics) [inline tests/test_autoplacement_qpcr_384.py]: Same as above

File: scripts/generate_hamilton_input_seq_quant_plate.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+
+        # assemble a dictionary that only contains PerInput outputs for each input based on input_output_maps
+
+        input_output_maps = self.process.input_output_maps
+        input_output_perinput_dict = {}
+
+        for input_output in input_output_maps:
+            input = input_output[0]['uri']
```
Comment (mwhamgenomics) [inline scripts/generate_hamilton_input_seq_quant_plate.py]: `input` is a builtin/reserved word in Python - would recommend using a different variable name

File: scripts/generate_hamilton_input_seq_quant_plate.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+            if input_output[1]['output-generation-type'] == 'PerInput':
+                if not input in input_output_perinput_dict.keys():
+                    input_output_perinput_dict[input] = [input_output[1]['uri']]
+                else:
+                    input_output_perinput_dict[input].append(input_output[1]['uri'])
+
+
+        # find all the inputs for the step that are analytes (i.e. samples and not associated files)
+        for input in input_output_perinput_dict:
```
Comment (mwhamgenomics) [inline scripts/generate_hamilton_input_seq_quant_plate.py]: Same as above

File: scripts/generate_hamilton_input_seq_quant_plate.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+            outputs = input_output_perinput_dict[input]
+
+            # the step requires 3 output replicates per input
+            if len(outputs) != 3:
+                raise InvalidStepError(message="3 replicates required for each sample and standard. Did you remember to click 'Apply' when assigning replicates?")
+
+
+            # remove semi-colon from input location as this is not compatible with Hamilton Venus software
+            input_location = input.location[1].replace(':', '')
```
Comment (mwhamgenomics) [inline scripts/generate_hamilton_input_seq_quant_plate.py]: The comment mentions removing semicolons, but this removes colons?

File: scripts/check_container_name.py [hunk 7]
```diff
@@ -1,39 +1,49 @@
...
+    script assumes that each suffix is only used once and that the suffixes are applied to the output containers in
+    the same order as they appear in the suffix argument
     """
-
+    _use_load_config = False  # prevent the loading of the config file
     def __init__(self, argv=None):
         # extra suffix argument required in addition to standard arg parser arguments
         super().__init__(argv)
         self.suffix = self.cmd_args.suffix
+        self.suffix = self.cmd_args.suffix
```
Comment (mwhamgenomics) [inline scripts/check_container_name.py]: `self.suffix` is being assigned twice

File: scripts/check_container_name.py [hunk 8]
```diff
@@ -1,39 +1,49 @@
...
         containers = self.process.output_containers()
 
+        suffixes = self.suffix
+
+
         for container in containers:
-            if not re.match(name_template, container.name):
-                print('%s is not a valid container name. Container names must match %s' % (container.name, name_template))
-                sys.exit(1)
+            valid_container='False'
```
Comment (mwhamgenomics) [inline scripts/check_container_name.py]: Can we use booleans for `valid_container`?

File: scripts/check_container_name.py [hunk 9]
```diff
@@ -27,14 +30,26 @@ def _run(self):
...
-            if not re.match(name_template, container.name):
-                print('%s is not a valid container name. Container names must match %s' % (container.name, name_template))
-                sys.exit(1)
+        suffixes = self.suffix
+        #suffixes.append(self.suffix)
+        if len(suffixes) != len(containers):
+            raise InvalidStepError(
+                message="The number of plate name suffixes must match the number of output containers. %s plate"
+                        "name suffixes configured for this step and %s output containers present. The expected suffixes are %s."
+                        % (str(len(suffixes)), str(len(containers)), str(suffixes)))
```
Comment (mwhamgenomics) [inline scripts/check_container_name.py]: `str` casts can be removed

File: scripts/qc_check.py [hunk 10]
```diff
@@ -55,11 +57,19 @@ def add_args(argparser):
                                help='output/input  QC udfs that should be updated as a result of the comparison')
         argparser.add_argument('-w', '--qc_results', nargs='*',
                                help='entries that should go into the output QC UDFs if a fail')
+        argparser.add_argument('-ps', '--passed', type=str, required=False,
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/qc_check.py]: changed to -d left --passed

File: scripts/next_step_assignment_kapa_qc.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+from pyclarity_lims.entities import Protocol
+
+
+class AssignNextStep(StepEPP):
+    _use_load_config = False  # prevent the loading of the config file
+    """
+    This script checks the KAPA Next Step udf and assigns the input sample to the next step in the protocol or to remove from workflow 
+    """
+    def __init__(self, argv=None):
+        super().__init__(argv)
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/next_step_assignment_kapa_qc.py]: removed

File: scripts/qc_check.py [hunk 12]
```diff
@@ -55,11 +57,19 @@ def add_args(argparser):
...
+                               help='entries that should go into the output QC UDFs if passes')
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/qc_check.py]: No, one value only.

File: scripts/autoplacement_qpcr_384.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+
+# Script for performing autoplacement of samples. If 1 input and 1 output then 1:1 placement. If multiple input plates then
+# takes all samples from each plate before next plate Loops through all inputs and assigns them to the next available space
+# by column-row in the output plate
+class AutoplacementQPCR384(StepEPP):
+    _use_load_config = False  # prevent the loading of the config
+
+    def __init__(self, argv=None):
+        super().__init__(argv)
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_qpcr_384.py]: removed

File: scripts/autoplacement_qpcr_384.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        input_container_nested_dict = {}
+
+        # update of container requires list variable containing the containers, only one container will be present in step
+        # because the container has not yet been fully populated then it must be obtained from the step rather than output
+        output_container_list = self.process.step.placements.get_selected_containers()
+
+        standards_dict = {}
+        outputs_dict = {}
+
+        for input in all_inputs:
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_qpcr_384.py]: refactored

File: scripts/autoplacement_qpcr_384.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+                for output in outputs:
+                    standards_dict[str(input.name) + str(output_counter)] = output
+                    output_counter += 1
+
+            elif input.name.split(" ")[0] == "No":
+
+                for output in outputs:
+                    # want no template controles to appear after all standards in the sorted dictionary
+                    standards_dict[str("z") + str(output_counter)] = output
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_qpcr_384.py]: casting removed

File: scripts/autoplacement_seq_plate_quant.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+from EPPs.common import StepEPP, InvalidStepError
+
+
+# Script for performing autoplacement of samples for the Sequencing Plate Quantification step. Places standards SDNA Std A1 to H1
+# with 3 replicates in the first 3 columns (1 set of replicates per column). The remaining columns are filled with between 1 to 24
+# samples with 3 replicates. Replicates are placed across columns to allow for maximum efficiency when multiple samples taken from a
+# single input plate.
+
+class Autoplacement_seq_plate_quant(StepEPP):
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_seq_plate_quant.py]: refactored

File: scripts/autoplacement_seq_plate_quant.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    _use_load_config = False  # prevent the loading of the config
+
+    def __init__(self, argv=None):
+        super().__init__(argv)
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_seq_plate_quant.py]: Removed

File: scripts/autoplacement_seq_plate_quant.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        outputs_dict = {}
+
+        # update of container requires list variable containing the containers, only one container will be present in step
+        # because the container has not yet been fully populated then it must be obtained from the step rather than output
+        output_container_list = self.process.step.placements.get_selected_containers()
+
+        input_output_maps = self.process.input_output_maps
+
+        # assemble a dictionary that only contains PerInput outputs for each input
+        input_output_PerInput_dict = {}
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_seq_plate_quant.py]: variable nolonger used

File: scripts/autoplacement_seq_plate_quant.py [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+        for input_output in input_output_maps:
+            input = input_output[0]['uri']
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/autoplacement_seq_plate_quant.py]: refactored

File: scripts/check_container_name.py (hunk 7 above)
Comment (EGCGDanForth) [reply scripts/check_container_name.py]: corrected

File: scripts/check_container_name.py (hunk 8 above)
Comment (EGCGDanForth) [reply scripts/check_container_name.py]: changed to boolean

File: scripts/generate_hamilton_input_seq_quant_plate.py (hunk 4 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_seq_quant_plate.py]: refactored

File: scripts/generate_hamilton_input_seq_quant_plate.py (hunk 5 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_seq_quant_plate.py]: line no longer present

File: scripts/generate_hamilton_input_seq_quant_plate.py (hunk 6 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_seq_quant_plate.py]: comment updated

File: scripts/generate_hamilton_input_uct.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    _use_load_config = False  # prevent the loading of the config
+    csv_column_headers = []
+    output_file_name = 'KAPA_MAKE_LIBRARIES.csv'
+
+    def _generate_csv_dict(self):
+        # csv_dict will be a dictionary that consists of the lines to be present in the Hamilton input file.
+        csv_dict = {}
+
+        # find all the inputs for the step that are analytes (i.e. samples and not associated files)
+        for input in self.artifacts:
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_uct.py]: refactored

File: scripts/next_step_assignment_kapa_qpcr.py [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+class AssignNextStepKAPAqPCR(StepEPP):
+    """
+    This script assigns the next step for samples based on the QC flag of the input artifact and the step UDF
+    standard curve result.
+    """
+    _use_load_config = False  # prevent the loading of the config
+
+    def __init__(self, argv=None):
+        super().__init__(argv)
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/next_step_assignment_kapa_qpcr.py]: removed

File: scripts/normalisation.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class CalculateVolumes(StepEPP):
+    """
+    Calcuate the volume of sample and volume of buffer required based on step UDFs determining a target concentration
+    and overall volume. Arguments are used to specify the UDFs involved. Volumes are rounded to 1dp.
+    """
+    _use_load_config = False  # prevent the loading of the config
+
+    # use arguments to determing
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/normalisation.py]: comment removed

File: scripts/normalisation.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _run(self):
+        # obtain the target volume and concentration from the step UDFs specified by the arguments
+        target_volume = self.process.udf.get(self.target_volume_udf)
+        target_concentration = self.process.udf.get(self.target_conc_udf)
+        # create the set to hold the inputs to be updated with the output form the calculation
+        artifacts_to_update = set()
+
+        # obtain the concentration of each input and use that calculate the volume of sample and buffer required
+        for input in self.process.all_inputs():
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/normalisation.py]: refactored

File: tests/test_autoplacement_qpcr_384.py (hunk 3 above)
Comment (EGCGDanForth) [reply tests/test_autoplacement_qpcr_384.py]: removed

File: tests/test_generate_hamilton_input_cfp.py [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+    def fake_all_outputs(unique=False, resolve=False):
+        return [o for sublist in outputs.values() for o in sublist]
+    return fake_all_outputs
+
+def get_fake_outputs_per_input(outputs):
+    def fake_outputs_per_input(inputid, Analyte=False):
+        return outputs[inputid]
+    return fake_outputs_per_input
+
+def fake_outputs_per_input2(inputid, Analyte=False):
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply tests/test_generate_hamilton_input_cfp.py]: removed

File: tests/test_generate_hamilton_input_cfp.py [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+            GenerateHamiltonInputCFP,
+            'process', new_callable=PropertyMock(
+                return_value=Mock(all_inputs=fake_all_inputs1,
+                                  all_outputs=get_fake_all_outputs(outputs3),
+                                  outputs_per_input=get_fake_outputs_per_input(outputs3),
+                                  step=dummystep)
+            )
+        )
+
+        # self.patched_process4 = patch.object(
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply tests/test_generate_hamilton_input_cfp.py]: remove

File: tests/test_generate_hamilton_input_ntp.py [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+    return fake_all_outputs
+
+
+def get_fake_outputs_per_input(outputs):
+    def fake_outputs_per_input(inputid, Analyte=False):
+        return outputs[inputid]
+    return fake_outputs_per_input
+
+
+def fake_reagent_lots():
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply tests/test_generate_hamilton_input_ntp.py]: removed

File: tests/test_normalisation.py [hunk 27]
```diff
@@ -0,0 +1,56 @@
+from unittest.mock import Mock, patch, PropertyMock
+
+from scripts.normalisation import CalculateVolumes
+from tests.test_common import TestEPP
+
+
+def fake_outputs_per_input(inputid, Analyte=False):
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply tests/test_normalisation.py]: removed

File: scripts/generate_hamilton_input_uct.py [hunk 28]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python
+import csv
+import sys
+
+from EPPs.common import StepEPP, GenerateHamiltonInputEPP, InvalidStepError
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_uct.py]: no longer applicable

File: tests/test_generate_hamilton_input_cfp.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (mwhamgenomics) [reply tests/test_generate_hamilton_input_cfp.py]: This and `fake_outputs_per_input3` are still present and unused.

File: scripts/assign_workflow_kapa_qc.py [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+
+
+    @staticmethod
+    def add_args(argparser):
+        argparser.add_argument('-sw', '--sppw', type=str, required=True, help='Sequencing plate prepration workflow name')
+        argparser.add_argument('-ss', '--spps',  type=str, required=True, help='Sequencing plate preparation step name')
+        argparser.add_argument('-rw', '--rrw', type=str, required=True, help='Request repeats workflow name')
+        argparser.add_argument('-rs', '--rrs',  type=str, required=True, help='Request repeats step name')
+        argparser.add_argument('-qw', '--qpcrw', type=str, required=True, help='Make and Read QPCR workflow name')
+        argparser.add_argument('-qs', '--qpcrs',  type=str, required=True, help='Make and Read QPCR step name')
```
Comment (mwhamgenomics) [inline scripts/assign_workflow_kapa_qc.py]: Argparse is not designed to handle single-dash options more than one character long.

File: EPPs/common.py [hunk 30]
```diff
@@ -123,7 +130,10 @@ def _run(self):
 
     def run(self):
         try:
-            self._run()
+            return self._run()
+        except InvalidStepError as e:
+            print(e.message)
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (tcezard) [reply EPPs/common.py]: I changed the `BaseException` to `Exception`. 
Indeed the goal was to have a single message that can be displayed on the LIMS step

File: scripts/generate_hamilton_input_cfp.py [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+        RSB_template = "LP[0-9]{7}-RSB"
+        reagent_lots = list(self.process.step.reagent_lots)
+
+        rsb_barcode = None
+        for lot in reagent_lots:
+            if re.match(RSB_template, lot.lot_number):
+                rsb_barcode = lot.lot_number
+
+        if not rsb_barcode:
+            raise InvalidStepError('Please assign RSB lot before generating Hamilton input.')
```
Comment (mwhamgenomics) [inline scripts/generate_hamilton_input_cfp.py]: Lines 28-38 are duplicated in generate_hamilton_input_imp_ssqc and generate_hamilton_input_ntp - should be merged

File: scripts/assign_workflow_kapa_qc.py (hunk 29 above)
Comment (EGCGDanForth) [reply scripts/assign_workflow_kapa_qc.py]: I have changed the single-dash alternatives to arbitrary single character values but the two character values above were working in testing in the system.

File: scripts/email_sample_disposal_review.py [hunk 32]
```diff
@@ -7,8 +7,6 @@
 class SampleDisposalFMEmail(SendMailEPP):
 
     def _run(self):
-        if len(self.projects) > 1:  # check if more than one project in step, only one permitted
-            raise ValueError('More than one project present in step. Only one project per step permitted')
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/email_sample_disposal_review.py]: Removed (although not intended to be part of this branch)

File: scripts/generate_hamilton_input_cfp.py (hunk 31 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_cfp.py]: New function added to common.py and duplicated lines removed from scripts

File: scripts/next_step_assignment_seq_pico.py [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+class AssignNextStepSeqPico(StepEPP):
+    _use_load_config = False  # prevent the loading of the config
+
+    """
+    This script assigns the next step for samples in the Sequencing Plate Picgreen step to either "review" if they
+    have failed or complete (as last step in the protocol).
+    """
+
+    def __init__(self, argv=None):
+        super().__init__(argv)
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/next_step_assignment_seq_pico.py]: removed

File: scripts/subtraction.py [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+    # the -n argument can have as many entries as required, if there are spaces within the udfnames and when running locally
+    # then they must be enclosed in speech marks " not quotes ' -n "udf name 1" "udf name 2" "udfname 3" BUT use quotes ' when configuring the EPP as speech marks close the
+    # bash command e.g.
+
+    # Parse command line options
+    args = p.parse_args()
+
+    # Setup the EPP
+    action = VolumeUpdate(
+        args.step_uri, args.username, args.password, args.step_udf_names, args.sample_udf_names, args.log_file,
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/subtraction.py]: Arg parsing updated (will also add test)

File: tests/test_spectramax.py [hunk 35]
```diff
@@ -19,7 +20,15 @@ class TestSpectramaxOutput(TestEPP):
 
     exp_plate_names = ['a_plate', 'another_plate', 'yet_another_plate']
 
+
+
+
     def setUp(self):
+
+        self.patched_object = patch.object(
```
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply tests/test_spectramax.py]: Removed

File: scripts/subtraction.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (EGCGDanForth) [reply scripts/subtraction.py]: Script not required so removed from branch

**Pull request merged**